### PR TITLE
Refactor computing build results

### PR DESCRIPTION
**Pull request by Martchus**
First step to implement meta-groups for products: https://progress.opensuse.org/issues/13742

combined_build_results

### CODE REVIEW

**Comment by coolo** [comment]
you better use real life data for such experiments. Pick one of /work/users/coolo/SQL-DUMPS/

**Comment by coolo** [comment]
I accept this as experiment, but hopefully we agree that this hack has no business in git. This is not a meta group - this is a if cascade.

**Comment by Martchus** [comment]
I agree that this approach is a hack. It is just a WIP commit and I opened the pull request to receive feedback, not for this being merged now.

If you have a better idea to query common builds to show combined results, please share your ideas. I can think of using a GROUP BY database query but I don't know how that would be done using DBIx.

If you totally disagree with the filtering approach, I can follow the 2nd implementation proposal mentioned in the ticked.

**Comment by coolo** [comment]
I'm fine with the filtering approach, but you should do this OO - have a group filter its results and then have a meta group object that filters multiple results.

**Comment by Martchus** [comment]
Ok, sounds like the kind of re-factoring I had in mind anyways but no fundamental change in the algorithm itself.
But maybe we could improve here, too. You're very experienced with the database and the abstraction framework openQA uses, so do you think it would be possible to let the database do the grouping (eg. by using `GROUP BY` -SQL-Syntax). This would be likely more efficient and less error prone.

**Comment by coolo** [comment]
I don't really understand what you have in mind, so better consult with me in person about that ;)

File: lib/OpenQA/Schema.pm [hunk 1]
```diff
@@ -26,7 +26,7 @@ use FindBin qw/$Bin/;
 
 # after bumping the version please look at the instructions in the docs/Contributing.asciidoc file
 # on what scripts should be run and how
-our $VERSION   = 45;
+our $VERSION   = 46;
```
Comment (coolo) [inline lib/OpenQA/Schema.pm]: you didn't change the database schema

**Comment by Martchus** [comment]
See the commit message for the current state. I think it is already useful combined with the `group` parameter and will not harm anybody as it is completely optional.

Considering other criteria beside the build ID such as distri, version and arch would be the next step.

File: lib/OpenQA/WebAPI/Controller/Main.pm [hunk 2]
```diff
@@ -20,130 +20,46 @@ use warnings;
...
-        if (%$res) {
-            $res->{_group} = $group;
-            push(@results, $res);
-        }
+        my $build_results = OpenQA::BuildResults::compute_build_results($group, $limit_builds, $time_limit_days);
+        push(@results, $build_results) if $build_results;
     }
-    $self->stash('results', \@results);
+
+    my $meta_groups
```
Comment (coolo) [inline lib/OpenQA/WebAPI/Controller/Main.pm]: This is just obscure. Declare the variable (which will be undef by default) and then assign it and use if for that.

**Comment by coolo** [comment]
Please add a test case verifying groups not having seen jobs in 14 days are not displayed on the dashboard. Neither with or without merge_builds

File: lib/OpenQA/BuildResults.pm [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+                || $job->state eq OpenQA::Schema::Result::Jobs::OBSOLETED)
+            {
+                next;        # ignore
+            }
+            my $state = $job->state;
+            if (grep { /$state/ } (OpenQA::Schema::Result::Jobs::PENDING_STATES)) {
+                $jr{inprogress}++;
+                next;
+            }
+            #$self->app->log->error("MISSING S:" . $job->state . " R:" . $job->result);
```
Comment (Martchus) [inline lib/OpenQA/BuildResults.pm]: FIXME

**Comment by coolo** [comment]
I think your PR just shows that this approach doesn't work as naively. https://w3.nue.suse.com/~coolo/screencapture-gertrude-suse-de-1474888542192.png is in no way more presenting the product.

**Comment by okurz** [comment]
https://w3.nue.suse.com/~okurz/openQA_group_and_merge_builds.png looks pretty ok although the first combined group is not really helpful. When you not filter the groups certainly some combined groups don't have a useful meaning.

**Comment by coolo** [comment]
that's a very carefully crafted URL - respect :)

**Comment by okurz** [comment]
that's why we have dark launched the feature after all and added UI elements 
where we could provide only useful sets

File: lib/OpenQA/BuildResults.pm [hunk 4]
```diff
@@ -0,0 +1,206 @@
+# Copyright (C) 2016 SUSE Linux Products GmbH
```
Comment (aaannz) [inline lib/OpenQA/BuildResults.pm]: We have a different company for IP. Use SUSE LLC.

File: lib/OpenQA/BuildResults.pm [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+package OpenQA::BuildResults;
+use OpenQA::Schema::Result::Jobs;
+use Date::Format;
+use strict;
```
Comment (aaannz) [inline lib/OpenQA/BuildResults.pm]: This IMO should be first thing after `package`. Perl critic complains on this on stricter levels.

**Comment by Martchus** [comment]
Since we came to the conclusion that we should have well defined meta-groups to accumulate build results for products I moved this approach to another branch and keep only the re-factoring part here.

I can still implement the test for the `time_limit_days` parameter. However, I would do it in an extra commit/PR since it is actually independent of the meta groups feature.

File: templates/main/index.html.ep [hunk 6]
```diff
@@ -21,8 +21,6 @@
 
 % for my $groupresults (@$results) {
     % my $group = delete $groupresults->{_group};
-    <h2>
-        %= link_to $group->name => url_for('group_overview', groupid => $group->id)
-    </h2>
+    <h2><%= link_to $group->name => url_for('group_overview', groupid => $group->id) %></h2>
```
Comment (coolo) [inline templates/main/index.html.ep]: I wonder why you changed this. I prefer the %= syntax over the %></ inline smileys

**Comment by coolo** [comment]
I get this: 

```
coolo@gertrude#openQA>[Mon Sep 26 16:25:30 2016] [19032:debug] GET "/"
[Mon Sep 26 16:25:30 2016] [19032:debug] Routing to controller "OpenQA::WebAPI::Controller::Main" and action "index"
[Mon Sep 26 16:25:30 2016] [19032:debug] Retrieving results for up to 3 builds up to 14 days old
[Mon Sep 26 16:25:30 2016] [19032:debug] [DBIx debug] Took 0.00058699 seconds executed: SELECT me.id, me.name, me.size_limit_gb, me.keep_logs_in_days, me.t_created, me.t_updated FROM job_groups me ORDER BY name: .
[Mon Sep 26 16:25:30 2016] [19032:error] DBIx::Class::Storage::DBI::_dbh_execute(): DBI Exception: DBD::Pg::st execute failed: FEHLER:  Zeitzonenunterschied ist außerhalb des gültigen Bereichs: „-244395-02-08 14:25:30“ [for Statement "SELECT me.BUILD, MIN( t_created ) AS first_hit FROM jobs me WHERE ( ( me.group_id = ? AND me.t_created > ? ) ) GROUP BY BUILD ORDER BY first_hit DESC" with ParamValues: 1='25', 2='-244395-02-08 14:25:30'] at /space/prod/openQA/script/../lib/OpenQA/BuildResults.pm line 38

[Mon Sep 26 16:25:30 2016] [19032:debug] Template "exception.development.html.ep" not found
[Mon Sep 26 16:25:30 2016] [19032:debug] Rendering cached template "exception.html.ep"
[Mon Sep 26 16:25:30 2016] [19032:debug] Rendering cached template "layouts/error.html.ep"
[Mon Sep 26 16:25:30 2016] [19032:debug] 500 Internal Server Error (0.014903s, 67.101/s)
```

File: templates/main/index.html.ep (hunk 6 above)
Comment (Martchus) [inline templates/main/index.html.ep]: Ok

**Comment by Martchus** [comment]
I only tested with sqlite and it works. When using postgresql now I'm getting the same error so I'll check what's the cause.

**Comment by coolo** [comment]
We should get rid of sqlite *now* - this dual DB madness is a problem ;(

**Comment by Martchus** [comment]
The issue was just caused because I set `$time_limit_days` to 90000000 for testing the test for it. Seems like PostgeSQL can not deal with it.

**Comment by coolo** [comment]
can you? :)

**Comment by coolo** [comment]
but that makes my wanted test case even more necessary - you should verify time_limit_days=1 shows less than time_limit_days=200 :)

**Comment by Martchus** [comment]
I was able to reproduce the failing check after 19 iterations on master - however, it failed in another line. I'll check the tests tomorrow.

File: lib/OpenQA/WebAPI/Controller/Main.pm [hunk 7]
```diff
@@ -20,130 +20,38 @@ use warnings;
...
 
 sub group_overview {
     my ($self) = @_;
 
-    my $limit_builds = $self->param('limit_builds') // 10;
-    my $only_tagged  = $self->param('only_tagged')  // 0;
-    my $group        = $self->db->resultset('JobGroups')->find($self->param('groupid'));
+    my $limit_builds    = $self->param('limit_builds')    // 10;
+    my $time_limit_days = $self->param('time_limit_days') // 14;
+    $self->app->log->debug("Retrieving results for up to $limit_builds builds up to $time_limit_days days old");
```
Comment (okurz) [inline lib/OpenQA/WebAPI/Controller/Main.pm]: you are duplicating this line and the previous one so I consider this a very bad refactoring

File: lib/OpenQA/BuildResults.pm [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+package OpenQA::BuildResults;
+use strict;
+use OpenQA::Schema::Result::Jobs;
+use Date::Format;
+
+sub compute_build_results {
+    my ($app, $group, $limit, $time_limit_days) = @_;
```
Comment (okurz) [inline lib/OpenQA/BuildResults.pm]: $app is only used for logging so can you try to forward only the logger object?

File: lib/OpenQA/WebAPI/Controller/Main.pm (hunk 7 above)
Comment (Martchus) [inline lib/OpenQA/WebAPI/Controller/Main.pm]: It depends on whether the default values will always be the same in both cases or might differ in the future. I wasn't sure so I've kept them separate.

File: lib/OpenQA/BuildResults.pm (hunk 8 above)
Comment (Martchus) [inline lib/OpenQA/BuildResults.pm]: I could improve that when touching the code anyways.

Comment (coolo) [inline lib/OpenQA/BuildResults.pm]: This is not necessary anyway as openQA::Utils provides log_debug for such cases.

**Comment by okurz** [comment]
yes, that would be the even better choice

**Comment by Martchus** [comment]
Ok, so I'll using this method when touching the code anyways.

**Pull request merged**